File: 5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        int start =0, end = 0;
        for(int i = 0; i< s.length(); i++) {
           int len1 = expandFromCenter(s, i, i+1);
           int len2 = expandFromCenter(s, i, i);
           int maxLen = Math.max(len1, len2);
           if(maxLen > end - start) {
            start = i - (maxLen - 1) / 2 ;
            end = i + maxLen / 2;
           }
        }
        return s.substring(start, end + 1);
    }
    
    private int expandFromCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        // -1 3
        return right - left - 1;
    }
}
// @lc code=end

File: js/5.longest-palindromic-substring.js
/*
 * @lc app=leetcode id=5 lang=javascript
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
    let maxLength = 0;
    let maxString = "";
    function isPalindromic(start, end) {
        while (start >= 0 && end < s.length) {
            if (s[start] !== s[end]) break;
            else {
                if (end - start + 1 >= maxLength) {
                    maxLength = end - start + 1;
                    maxString = s.slice(start, end + 1);
                }
                start--;
                end++;
            }
        }
    }
    for (let i = 0; i < s.length; i++) {
        isPalindromic(i, i);
        isPalindromic(i, i + 1);
    }
    return maxString;

};
// @lc code=end

